google-cloud-spanner-executor/src/main/java/com/google/cloud/executor/spanner/CloudClientExecutor.java

```diff
@@ -0,0 +1,3482 @@
+    }
+
+    com.google.cloud.spanner.Struct.Builder builder = com.google.cloud.spanner.Struct.newBuilder();
+    for (int i = 0; i < fieldTypes.size(); ++i) {
+      builder
+          .set(fieldTypes.get(i).getName())
+          .to(valueProtoToCloudValue(fieldTypes.get(i).getType(), fieldValues.get(i)));
+    }
+    return builder.build();
+  }
+
+  /** Convert a Type proto to a cloud Type. */
+  private static com.google.cloud.spanner.Type typeProtoToCloudType(
+      com.google.spanner.v1.Type typeProto) {
+    switch (typeProto.getCode()) {
+      case BOOL:
+        return com.google.cloud.spanner.Type.bool();
+      case INT64:
+        return com.google.cloud.spanner.Type.int64();
+      case STRING:
+        return com.google.cloud.spanner.Type.string();
+      case BYTES:
+        return com.google.cloud.spanner.Type.bytes();
+      case FLOAT64:
+        return com.google.cloud.spanner.Type.float64();
+      case DATE:
+        return com.google.cloud.spanner.Type.date();
+      case TIMESTAMP:
+        return com.google.cloud.spanner.Type.timestamp();
+      case NUMERIC:
+        if (typeProto.getTypeAnnotation().equals(TypeAnnotationCode.PG_NUMERIC)) {
+          return com.google.cloud.spanner.Type.pgNumeric();
+        } else {
+          return com.google.cloud.spanner.Type.numeric();
+        }
+      case STRUCT:
+        List<StructType.Field> fields = typeProto.getStructType().getFieldsList();
+        List<com.google.cloud.spanner.Type.StructField> cloudFields = new ArrayList<>();
+        for (StructType.Field field : fields) {
+          com.google.cloud.spanner.Type fieldType = typeProtoToCloudType(field.getType());
+          cloudFields.add(com.google.cloud.spanner.Type.StructField.of(field.getName(), fieldType));
+        }
+        return com.google.cloud.spanner.Type.struct(cloudFields);
+      case ARRAY:
+        com.google.spanner.v1.Type elementType = typeProto.getArrayElementType();
+        if (elementType.getCode() == TypeCode.ARRAY) {
+          throw SpannerExceptionFactory.newSpannerException(
+              ErrorCode.INVALID_ARGUMENT, "Unsupported array-of-array proto type");
+        } else {
+          com.google.cloud.spanner.Type cloudElementType = typeProtoToCloudType(elementType);
+          return com.google.cloud.spanner.Type.array(cloudElementType);
+        }
+      case JSON:
+        if (typeProto.getTypeAnnotation().equals(TypeAnnotationCode.PG_JSONB)) {
+          return com.google.cloud.spanner.Type.pgJsonb();
+        } else {
+          return com.google.cloud.spanner.Type.json();
+        }
+      default:
+        throw SpannerExceptionFactory.newSpannerException(
+            ErrorCode.INVALID_ARGUMENT, "Unsupported proto type: " + typeProto);
+    }
+  }
+
+  /** Convert a cloud Type to a Type proto. */
+  private static com.google.spanner.v1.Type cloudTypeToTypeProto(@NotNull Type cloudTypeProto) {
```

Comment:
Can we re-use https://github.com/googleapis/java-spanner/blob/main/google-cloud-spanner/src/main/java/com/google/cloud/spanner/Type.java#L427?

Reply:
This API is not public.